File: web/dr-viewer/analyze/fmlIssueParser.js
/**
 * Port of FmlIssueParser.java — NeoForge multi-block + dependency/conflict banners (G-10 / CA-18).
 */

const ISSUE_HEADER_RE = /^--\s*Mod loading issue(?:\s+for:\s*([^-\n]+))?\s*--\s*$/i;
const MOD_ID_RE = /^\s*Mod ID:\s*(.+)\s*$/i;
const MOD_FILE_RE = /^\s*Mod File:\s*(.+)\s*$/i;
const FAILURE_RE = /^\s*Failure message:\s*(.+)\s*$/i;
const KIND_RE = /^\s*Issue kind:\s*(.+)\s*$/i;
const JAR_NAME_RE = /([\w.+-]+\.jar)/gi;

const KIND_RANK = {
  mod_load_dependency: 0,
  mod_corrupt: 1,
  mod_load_failed: 2,
  mod_load_script: 3,
};

const BANNER_KINDS = {
  missing_unsupported: {
    id: 'fml_missing_unsupported_dependencies',
    priority: 100,
    messageKey: 'fml.missing_unsupported_dependencies',
    issueKind: 'mod_load_dependency',
  },
  conflicts: {
    id: 'fml_conflicts_between_mods',
    priority: 90,
    messageKey: 'fml.conflicts_between_mods',
    issueKind: 'mod_load_dependency',
  },
  preload_incompat: {
    id: 'fml_preload_incompatible',
    priority: 85,
    messageKey: 'fml.preload_incompatible',
    issueKind: 'mod_load_dependency',
  },
  optional_unsupported: {
    id: 'fml_unsupported_optional_dependencies',
    priority: 40,
    messageKey: 'fml.unsupported_optional_dependencies',
    issueKind: 'mod_load_dependency',
  },
};

function stripJar(raw) {
  if (!raw) return null;
  let s = String(raw).trim();
  if (s.includes('/')) s = s.slice(s.lastIndexOf('/') + 1);
  if (s.includes('\\')) s = s.slice(s.lastIndexOf('\\') + 1);
  return s || null;
}

function inferKind(message, explicit) {
  if (explicit) {
    const k = explicit.trim().toLowerCase().replace(/\s+/g, '_');
    if (k.includes('depend')) return 'mod_load_dependency';
    if (k.includes('conflict') || k.includes('incompat')) return 'mod_load_dependency';
    if (k.includes('corrupt')) return 'mod_corrupt';
    if (k.includes('script')) return 'mod_load_script';
    if (k.includes('fail') || k.includes('load')) return 'mod_load_failed';
    return k;
  }
  const lower = (message || '').toLowerCase();
  if (lower.includes('missing dependency') || lower.includes('requires ')
    || lower.includes('mandatory depend') || lower.includes('conflicts between')
    || lower.includes('incompatibilit')) {
    return 'mod_load_dependency';
  }
  if (lower.includes('does not exist') || lower.includes('corrupt')) return 'mod_corrupt';
  return 'mod_load_failed';
}

function kindSortKey(kind) {
  return KIND_RANK[kind] ?? 9;
}

function detectBanner(line) {
  if (!line) return null;
  const lower = line.toLowerCase();
  if (lower.includes('missing or unsupported mandatory dependencies:')) {
    return BANNER_KINDS.missing_unsupported;
  }
  if (lower.includes('conflicts between mods:')
    || lower.includes('incompatibilities between mods:')) {
    return BANNER_KINDS.conflicts;
  }
  if (lower.includes('unsupported installed optional dependencies:')) {
    return BANNER_KINDS.optional_unsupported;
  }
  if (lower.includes('some of your mods are incompatible with the game or each other')) {
    return BANNER_KINDS.preload_incompat;
  }
  return null;
}

export function isMessageContinuation(line) {
  if (line == null || line === '') return false;
  const trimmed = line.trim();
  if (trimmed === 'More details:'
    || trimmed === 'A potential solution has been determined, this may resolve your problem:') {
    return true;
  }
  const first = line.charAt(0);
  if (first !== ' ' && first !== '\t') return false;
  if (trimmed.includes('Issues may arise. Continue at your own risk.')) return false;
  return true;
}

export function modIdsFromQuotedLabels(line) {
  const modIds = [];
  if (!line || !String(line).trim()) return modIds;
  const parts = String(line).split("'");
  for (let i = 0; i < parts.length - 1; i++) {
    const prefix = parts[i];
    if (prefix.endsWith('Mod ID: ')
      || prefix.endsWith('Requested by: ')
      || prefix.endsWith('Mod ')
      || prefix.endsWith('discourages ')) {
      const id = parts[i + 1].trim();
      if (id && id.toLowerCase() !== '[missing]') modIds.push(id);
    }
  }
  return modIds;
}

function jarNamesFromLine(line) {
  const jars = [];
  if (!line) return jars;
  JAR_NAME_RE.lastIndex = 0;
  let m;
  while ((m = JAR_NAME_RE.exec(line)) !== null) jars.push(m[1]);
  return jars;
}

/**
 * Known FML pattern hits for Issues / boot triage. Sorted by priority descending.
 */
export function parseKnownPatternHits(text) {
  if (!text || !String(text).trim()) return [];
  const hits = [];
  const lines = text.split(/\r?\n/);
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const kind = detectBanner(line);
    if (!kind || !line.includes(']: ')) continue;
    if (i + 1 >= lines.length || !isMessageContinuation(lines[i + 1])) continue;
    const modIds = new Set();
    const jarNames = new Set();
    for (let j = i + 1; j < lines.length; j++) {
      if (!isMessageContinuation(lines[j])) break;
      for (const id of modIdsFromQuotedLabels(lines[j])) modIds.add(id);
      for (const jar of jarNamesFromLine(lines[j])) jarNames.add(jar);
    }
    const row = {
      id: kind.id,
      priority: kind.priority,
      message_key: kind.messageKey,
      mod_ids: [...modIds],
    };
    if (jarNames.size) row.jar_names = [...jarNames];
    hits.push(row);
  }
  hits.sort((a, b) => b.priority - a.priority);
  return hits;
}

function parseBanners(text, issues) {
  const lines = text.split(/\r?\n/);
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const kind = detectBanner(line);
    if (!kind || !line.includes(']: ')) continue;
    if (i + 1 >= lines.length || !isMessageContinuation(lines[i + 1])) continue;
    const after = line.slice(line.indexOf(']: ') + 3).trim();
    const modIds = new Set();
    const jarNames = new Set();
    const msgParts = [after];
    for (let j = i + 1; j < lines.length; j++) {
      if (!isMessageContinuation(lines[j])) break;
      msgParts.push(lines[j].trimStart());
      for (const id of modIdsFromQuotedLabels(lines[j])) modIds.add(id);
      for (const jar of jarNamesFromLine(lines[j])) jarNames.add(jar);
    }
    const issue = {
      mod_id: modIds.size ? [...modIds][0] : 'unknown',
      kind: kind.issueKind,
      message: msgParts.join(' ').trim(),
      file: null,
      banner: true,
      priority: kind.priority,
      mod_ids: [...modIds],
      jar_names: [...jarNames],
    };
    issues.push(issue);
  }
}

/**
 * @param {string} text
 * @returns {object[]}
 */
export function parseFmlIssues(text) {
  if (!text) return [];

  const lines = text.split(/\r?\n/);
  const issues = [];
  let current = null;

  function flush() {
    if (!current) return;
    const modId = current.mod_id || current.header_mod || 'unknown';
    const message = current.message || 'Mod loading failed';
    const kind = inferKind(message, current.kind);
    const row = {
      mod_id: modId.trim(),
      kind,
      message: message.trim(),
      file: stripJar(current.file),
    };
    if (current.mod_ids?.size) row.mod_ids = [...current.mod_ids];
    if (current.jar_names?.size) row.jar_names = [...current.jar_names];
    issues.push(row);
    current = null;
  }

  for (const line of lines) {
    const header = ISSUE_HEADER_RE.exec(line.trim());
    if (header) {
      flush();
      current = {
        header_mod: header[1] ? header[1].trim() : null,
        mod_id: null,
        file: null,
        message: null,
        kind: null,
        mod_ids: new Set(),
        jar_names: new Set(),
      };
      if (current.header_mod) current.mod_ids.add(current.header_mod);
      continue;
    }
    if (!current) continue;

    if (line.trim().startsWith('-- ') && line.trim().endsWith(' --')) {
      flush();
      continue;
    }

    const modId = MOD_ID_RE.exec(line);
    if (modId) {
      current.mod_id = modId[1].trim();
      current.mod_ids.add(current.mod_id);
      continue;
    }
    const modFile = MOD_FILE_RE.exec(line);
    if (modFile) {
      current.file = modFile[1].trim();
      for (const jar of jarNamesFromLine(current.file)) current.jar_names.add(jar);
      continue;
    }
    const failure = FAILURE_RE.exec(line);
    if (failure) {
      current.message = failure[1].trim();
      continue;
    }
    const kind = KIND_RE.exec(line);
    if (kind) {
      current.kind = kind[1].trim();
      continue;
    }
    if (line.trim() && line.trim() !== 'Details:') {
      current.message = current.message
        ? `${current.message} ${line.trim()}`
        : line.trim();
    }
  }
  flush();

  parseBanners(text, issues);

  issues.sort((a, b) => kindSortKey(a.kind) - kindSortKey(b.kind)
    || String(a.mod_id).localeCompare(String(b.mod_id)));

  return issues.map((issue, i) => {
    const row = {
      rank: issue.banner ? (issue.priority >= 80 ? 1 : 2) : i + 1,
      mod_id: issue.mod_id,
      kind: issue.kind,
      message: issue.message,
      file: issue.file,
    };
    if (issue.mod_ids?.length) row.mod_ids = issue.mod_ids;
    if (issue.jar_names?.length) row.jar_names = issue.jar_names;
    if (issue.banner) {
      row.banner = true;
      row.priority = issue.priority;
    }
    return row;
  });
}

File: watchtower-core/src/main/java/dev/mcstatus/watchtower/core/collect/FmlIssueParser.java
package dev.mcstatus.watchtower.core.collect;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Parse NeoForge / FML {@code -- Mod loading issue --} blocks and dependency/conflict banners
 * from log or crash text (CA-18 / G-10).
 */
public final class FmlIssueParser {

    private static final Pattern HEADER = Pattern.compile(
            "^-- Mod loading issue(?: for: ([\\w.-]+))? --\\s*$");
    private static final Pattern MOD_ID = Pattern.compile(
            "^\\s*Mod ID:\\s*(.+?)\\s*$", Pattern.CASE_INSENSITIVE);
    private static final Pattern MOD_FILE = Pattern.compile(
            "^\\s*Mod File:\\s*(.+?)\\s*$", Pattern.CASE_INSENSITIVE);
    private static final Pattern FAILURE = Pattern.compile(
            "^\\s*Failure message:\\s*(.+?)\\s*$", Pattern.CASE_INSENSITIVE);
    private static final Pattern ISSUE_KIND = Pattern.compile(
            "^\\s*Issue kind:\\s*(.+?)\\s*$", Pattern.CASE_INSENSITIVE);
    private static final Pattern JAR_NAME = Pattern.compile(
            "([\\w.+-]+\\.jar)", Pattern.CASE_INSENSITIVE);

    private FmlIssueParser() {
    }

    public static JsonArray parse(String text) {
        JsonArray out = new JsonArray();
        if (text == null || text.isBlank()) {
            return out;
        }
        List<Issue> issues = new ArrayList<>();
        parseMultiblock(text, issues);
        parseBanners(text, issues);
        issues.sort(Comparator.comparingInt(Issue::rank));
        for (Issue issue : issues) {
            out.add(issueToJson(issue));
        }
        return out;
    }

    /**
     * Known FML pattern hits for Issues / boot triage. Sorted by priority descending.
     * Rank-1 (priority ≥ 80) may inform boot {@code mod_load_dependency}; lower ranks are Issues-only.
     */
    public static JsonArray parseKnownPatternHits(String text) {
        JsonArray out = new JsonArray();
        if (text == null || text.isBlank()) {
            return out;
        }
        List<PatternHit> hits = new ArrayList<>();
        String[] lines = text.split("\\R");
        for (int i = 0; i < lines.length; i++) {
            String line = lines[i];
            BannerKind kind = detectBanner(line);
            if (kind == null) {
                continue;
            }
            if (!line.contains("]: ")) {
                continue;
            }
            if (i + 1 >= lines.length || !isMessageContinuation(lines[i + 1])) {
                continue;
            }
            Set<String> modIds = new LinkedHashSet<>();
            Set<String> jarNames = new LinkedHashSet<>();
            for (int j = i + 1; j < lines.length; j++) {
                if (!isMessageContinuation(lines[j])) {
                    break;
                }
                modIds.addAll(modIdsFromQuotedLabels(lines[j]));
                jarNames.addAll(jarNamesFromLine(lines[j]));
            }
            PatternHit hit = new PatternHit();
            hit.id = kind.id;
            hit.priority = kind.priority;
            hit.messageKey = kind.messageKey;
            hit.modIds.addAll(modIds);
            hit.jarNames.addAll(jarNames);
            hits.add(hit);
        }
        hits.sort(Comparator.comparingInt((PatternHit h) -> h.priority).reversed());
        for (PatternHit hit : hits) {
            JsonObject row = new JsonObject();
            row.addProperty("id", hit.id);
            row.addProperty("priority", hit.priority);
            row.addProperty("message_key", hit.messageKey);
            JsonArray mods = new JsonArray();
            hit.modIds.forEach(mods::add);
            row.add("mod_ids", mods);
            if (!hit.jarNames.isEmpty()) {
                JsonArray jars = new JsonArray();
                hit.jarNames.forEach(jars::add);
                row.add("jar_names", jars);
            }
            out.add(row);
        }
        return out;
    }

    private static void parseMultiblock(String text, List<Issue> issues) {
        String[] lines = text.split("\\R");
        Issue current = null;
        StringBuilder messageBuf = new StringBuilder();
        for (String raw : lines) {
            String line = raw.stripTrailing();
            Matcher header = HEADER.matcher(line.strip());
            if (header.matches()) {
                if (current != null) {
                    finalizeIssue(current, messageBuf);
                    issues.add(current);
                }
                current = new Issue();
                if (header.group(1) != null && !header.group(1).isBlank()) {
                    current.modId = header.group(1).strip();
                    current.modIds.add(current.modId);
                }
                messageBuf.setLength(0);
                continue;
            }
            if (current == null) {
                continue;
            }
            if (line.strip().startsWith("-- ") && line.strip().endsWith(" --")) {
                finalizeIssue(current, messageBuf);
                issues.add(current);
                current = null;
                messageBuf.setLength(0);
                continue;
            }
            Matcher id = MOD_ID.matcher(line);
            if (id.matches()) {
                current.modId = id.group(1).strip();
                current.modIds.add(current.modId);
                continue;
            }
            Matcher file = MOD_FILE.matcher(line);
            if (file.matches()) {
                current.file = file.group(1).strip();
                current.jarNames.addAll(jarNamesFromLine(current.file));
                continue;
            }
            Matcher fail = FAILURE.matcher(line);
            if (fail.matches()) {
                current.message = fail.group(1).strip();
                continue;
            }
            Matcher kind = ISSUE_KIND.matcher(line);
            if (kind.matches()) {
                current.kind = normalizeKind(kind.group(1).strip());
                continue;
            }
            if (!line.isBlank() && !line.strip().equals("Details:")) {
                if (messageBuf.length() > 0) {
                    messageBuf.append(' ');
                }
                messageBuf.append(line.strip());
            }
        }
        if (current != null) {
            finalizeIssue(current, messageBuf);
            issues.add(current);
        }
    }

    private static void parseBanners(String text, List<Issue> issues) {
        String[] lines = text.split("\\R");
        for (int i = 0; i < lines.length; i++) {
            String line = lines[i];
            BannerKind kind = detectBanner(line);
            if (kind == null || !line.contains("]: ")) {
                continue;
            }
            if (i + 1 >= lines.length || !isMessageContinuation(lines[i + 1])) {
                continue;
            }
            Issue issue = new Issue();
            issue.kind = kind.issueKind;
            issue.banner = true;
            issue.priority = kind.priority;
            String after = line.substring(line.indexOf("]: ") + 3).strip();
            issue.message = after;
            StringBuilder msg = new StringBuilder(after);
            for (int j = i + 1; j < lines.length; j++) {
                if (!isMessageContinuation(lines[j])) {
                    break;
                }
                String cont = lines[j].stripLeading();
                if (msg.length() > 0) {
                    msg.append(' ');
                }
                msg.append(cont);
                issue.modIds.addAll(modIdsFromQuotedLabels(lines[j]));
                issue.jarNames.addAll(jarNamesFromLine(lines[j]));
            }
            issue.message = msg.toString().strip();
            if (!issue.modIds.isEmpty()) {
                issue.modId = issue.modIds.iterator().next();
            }
            issues.add(issue);
        }
    }

    private static BannerKind detectBanner(String line) {
        if (line == null) {
            return null;
        }
        String lower = line.toLowerCase(Locale.ROOT);
        if (lower.contains("missing or unsupported mandatory dependencies:")) {
            return BannerKind.MISSING_UNSUPPORTED;
        }
        if (lower.contains("conflicts between mods:")
                || lower.contains("incompatibilities between mods:")) {
            return BannerKind.CONFLICTS;
        }
        if (lower.contains("unsupported installed optional dependencies:")) {
            return BannerKind.OPTIONAL_UNSUPPORTED;
        }
        if (lower.contains("some of your mods are incompatible with the game or each other")) {
            return BannerKind.PRELOAD_INCOMPAT;
        }
        return null;
    }

    /**
     * Indented FML detail lines after a banner (and a few known non-timestamp continuations).
     */
    static boolean isMessageContinuation(String line) {
        if (line == null || line.isEmpty()) {
            return false;
        }
        String trimmed = line.strip();
        if (trimmed.equals("More details:")
                || trimmed.equals("A potential solution has been determined, this may resolve your problem:")) {
            return true;
        }
        char first = line.charAt(0);
        if (first != ' ' && first != '\t') {
            return false;
        }
        if (trimmed.contains("Issues may arise. Continue at your own risk.")) {
            return false;
        }
        return true;
    }

    static List<String> modIdsFromQuotedLabels(String line) {
        List<String> modIds = new ArrayList<>();
        if (line == null || line.isBlank()) {
            return modIds;
        }
        String[] parts = line.split("'");
        for (int i = 0; i < parts.length - 1; i++) {
            String prefix = parts[i];
            if (prefix.endsWith("Mod ID: ")
                    || prefix.endsWith("Requested by: ")
                    || prefix.endsWith("Mod ")
                    || prefix.endsWith("discourages ")) {
                String id = parts[i + 1].strip();
                if (!id.isEmpty() && !"[MISSING]".equalsIgnoreCase(id)) {
                    modIds.add(id);
                }
            }
        }
        return modIds;
    }

    private static List<String> jarNamesFromLine(String line) {
        List<String> jars = new ArrayList<>();
        if (line == null) {
            return jars;
        }
        Matcher m = JAR_NAME.matcher(line);
        while (m.find()) {
            jars.add(m.group(1));
        }
        return jars;
    }

    private static JsonObject issueToJson(Issue issue) {
        JsonObject row = new JsonObject();
        if (issue.modId != null) {
            row.addProperty("mod_id", issue.modId);
        }
        row.addProperty("kind", issue.kind != null ? issue.kind : "mod_load_failed");
        if (issue.message != null) {
            row.addProperty("message", issue.message);
        }
        if (issue.file != null) {
            row.addProperty("file", issue.file);
        }
        if (!issue.modIds.isEmpty()) {
            JsonArray arr = new JsonArray();
            issue.modIds.forEach(arr::add);
            row.add("mod_ids", arr);
        }
        if (!issue.jarNames.isEmpty()) {
            JsonArray arr = new JsonArray();
            issue.jarNames.forEach(arr::add);
            row.add("jar_names", arr);
        }
        if (issue.banner) {
            row.addProperty("banner", true);
            row.addProperty("priority", issue.priority);
            // Rank: priority ≥ 80 is boot-relevant (rank 1); lower is Issues-only (rank ≥ 2)
            row.addProperty("rank", issue.priority >= 80 ? 1 : 2);
        }
        return row;
    }

    private static void finalizeIssue(Issue issue, StringBuilder messageBuf) {
        if ((issue.message == null || issue.message.isBlank()) && messageBuf.length() > 0) {
            issue.message = messageBuf.toString().strip();
        }
        if (issue.kind == null || issue.kind.isBlank()) {
            issue.kind = inferKind(issue.message);
        } else {
            issue.kind = normalizeKind(issue.kind);
        }
        if (issue.modId != null && !issue.modId.isBlank()) {
            issue.modIds.add(issue.modId);
        }
    }

    private static String normalizeKind(String raw) {
        if (raw == null) {
            return "mod_load_failed";
        }
        String k = raw.strip().toLowerCase(Locale.ROOT).replace(' ', '_');
        if (k.contains("depend")) {
            return "mod_load_dependency";
        }
        if (k.contains("conflict") || k.contains("incompat")) {
            return "mod_load_dependency";
        }
        if (k.contains("corrupt")) {
            return "mod_corrupt";
        }
        if (k.contains("fail") || k.contains("load")) {
            return "mod_load_failed";
        }
        return k;
    }

    private static String inferKind(String message) {
        if (message == null) {
            return "mod_load_failed";
        }
        String lower = message.toLowerCase(Locale.ROOT);
        if (lower.contains("missing dependency") || lower.contains("requires")
                || lower.contains("mandatory depend") || lower.contains("conflicts between")
                || lower.contains("incompatibilit")) {
            return "mod_load_dependency";
        }
        if (lower.contains("corrupt") || lower.contains("does not exist")) {
            return "mod_corrupt";
        }
        return "mod_load_failed";
    }

    private enum BannerKind {
        MISSING_UNSUPPORTED("fml_missing_unsupported_dependencies", 100,
                "fml.missing_unsupported_dependencies", "mod_load_dependency"),
        CONFLICTS("fml_conflicts_between_mods", 90,
                "fml.conflicts_between_mods", "mod_load_dependency"),
        PRELOAD_INCOMPAT("fml_preload_incompatible", 85,
                "fml.preload_incompatible", "mod_load_dependency"),
        OPTIONAL_UNSUPPORTED("fml_unsupported_optional_dependencies", 40,
                "fml.unsupported_optional_dependencies", "mod_load_dependency");

        final String id;
        final int priority;
        final String messageKey;
        final String issueKind;

        BannerKind(String id, int priority, String messageKey, String issueKind) {
            this.id = id;
            this.priority = priority;
            this.messageKey = messageKey;
            this.issueKind = issueKind;
        }
    }

    private static final class PatternHit {
        String id;
        int priority;
        String messageKey;
        final Set<String> modIds = new LinkedHashSet<>();
        final Set<String> jarNames = new LinkedHashSet<>();
    }

    private static final class Issue {
        String modId;
        String kind;
        String message;
        String file;
        boolean banner;
        int priority;
        final Set<String> modIds = new LinkedHashSet<>();
        final Set<String> jarNames = new LinkedHashSet<>();

        int rank() {
            if ("mod_load_dependency".equals(kind)) {
                return 0;
            }
            if ("mod_corrupt".equals(kind)) {
                return 1;
            }
            return 2;
        }
    }
}
